hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/io/storage/TestHoodieSparkLanceReader.java

```diff
@@ -0,0 +1,513 @@
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static org.apache.hudi.io.storage.LanceTestUtils.createRow;
+import static org.apache.hudi.io.storage.LanceTestUtils.createRowWithMetaFields;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Tests for {@link HoodieSparkLanceReader}.
+ */
+public class TestHoodieSparkLanceReader {
+
+  @TempDir
+  File tempDir;
+
+  private HoodieStorage storage;
+  private SparkTaskContextSupplier taskContextSupplier;
+  private String instantTime;
+
+  @BeforeEach
+  public void setUp() throws IOException {
+    storage = HoodieTestUtils.getStorage(tempDir.getAbsolutePath());
+    taskContextSupplier = new SparkTaskContextSupplier();
+    instantTime = "20251201120000000";
+  }
+
+  @AfterEach
+  public void tearDown() throws Exception {
+    if (storage != null) {
+      storage.close();
+    }
+  }
+
+  @Test
+  public void testReadPrimitiveTypes() throws Exception {
+    // Create schema with primitive types
+    StructType schema = new StructType()
+        .add("id", DataTypes.IntegerType, false)
+        .add("name", DataTypes.StringType, true)
+        .add("age", DataTypes.LongType, true)
+        .add("score", DataTypes.DoubleType, true)
+        .add("active", DataTypes.BooleanType, true);
+
+    // Create test data
+    List<InternalRow> expectedRows = new ArrayList<>();
+    expectedRows.add(createRow(1, "Alice", 30L, 95.5, true));
+    expectedRows.add(createRow(2, "Bob", 25L, 87.3, false));
+    expectedRows.add(createRow(3, "Charlie", 35L, 92.1, true));
+
+    // Write and read back
+    StoragePath path = new StoragePath(tempDir.getAbsolutePath() + "/test_primitives.lance");
+    try (HoodieSparkLanceReader reader = writeAndCreateReader(path, schema, expectedRows)) {
+      // Verify record count
+      assertEquals(expectedRows.size(), reader.getTotalRecords(), "Record count should match");
+
+      // Verify schema
+      assertNotNull(reader.getSchema(), "Schema should not be null");
+
+      // Read all records
+      List<InternalRow> actualRows = readAllRows(reader);
+
+      // Verify record count
+      assertEquals(expectedRows.size(), actualRows.size(), "Should read same number of records");
+
+      // Verify each record
+      for (int i = 0; i < expectedRows.size(); i++) {
+        InternalRow expected = expectedRows.get(i);
+        InternalRow actual = actualRows.get(i);
+
+        assertEquals(expected.getInt(0), actual.getInt(0), "id field should match");
```

Comment:
Is there an issue when trying to directly assert the equality between the rows?

Reply:
If we can convert the unsafe row to the GenericInternalRow I think that will make it a lot easier as we add more types or extend these tests in the future.
If that's not possible, then this is fine

Reply:
addressed in recent commit.